pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java

```diff
@@ -585,20 +588,25 @@ private CompletableFuture<Long> individualAckWithTransaction(CommandAck ack) {
 
             totalAckCount.add(ackedCount);
         }
-
+        CompletableFuture<Void> ackTask = new CompletableFuture<>();
```

Comment:
Why not record the positions in the `on-going transaction`, so we can filter out the entries when sending them to clients.

Reply:
Then we can not wait for all pending ack tasks finish.

Reply:
Maybe we only need to create this future for individual ack mode.

Reply:
> Why not record the positions in the on-going transaction, so we can filter out the entries when sending them to clients.

+1

Reply:
There is `individualAckWithTransaction`.  `Only waiting transaction ack tasks instead of all pending ack tasks.

Reply:
>Maybe we only need to create this future for individual ack mode.
Good catch! Fixed.